### ISSUE TITLE
if eager evaluation

### ISSUE THREAD

**Issue by lgfausak** [birchb1024/yamp#6]
i hit some unexpected behavior with an if: then: else:.  Consider:
```
- define: {x: { one: 1, two: 2 } }
- defmacro:
    name: PRINTNUM
    args: [ data ]
    value: { if: { python_eval: 'data == "ONE"' }, then: x.one, else: x.two }
---
{ "output": { "test": { PRINTNUM: { data: "TWO" } } } }
```
running yamp:
```
$ yamp g.yaml 
output:
  test: 2
```
However, if the data is not there, then it gets an error:
```
- define: {x: { two: 2 } }
- defmacro:
    name: PRINTNUM
    args: [ data ]
    value: { if: { python_eval: 'data == "ONE"' }, then: x.one, else: x.two }
---
{ "output": { "test": { PRINTNUM: { data: "TWO" } } } }
```
and the output:
```
$ yamp g.yaml 
ERROR: /home/install/git/vcp2-inv/data/tpa-lab-1/g.yaml
<class '__main__.YampException'>
Subvariable "one" not found in x.one
```
It seems that both then: and else: are resolved before if: is tested

**Comment by birchb1024** [comment]
Thanks for identifying this one. 

This arises because Yamp is an inside-out "term rewriting" program. In order to evaluate it scans and substitutes _before_ it evaluates builtins like 'if'. Really 'if' is not a great name, it could have been called 'choose one'. But I like 'if' because we all recognise it. Anyhow, here is an even simpler reproduction of this effect:

```
if: false
then: env.not_there
``` 
This provokes an error because the dot notation `env.not_there` throws an error since `not_there` isn't in the environment. 

Usually when a variable is not defined Yamp simply expands with the variable itself. But I chose to throw an error in the dot notation. It's inconsistent. So I will remove the error. In future it will evaluate to `env.not_there` and your code will run as you expect.

**Comment by birchb1024** [comment]
Fixed in 0.3.3 - Errors in dot notation result in warning messages and original variable expansion instead of exception.

**Comment by birchb1024** [comment]
Hang on I could do this better. I will see if I an pre-fetch a map key and check if it's Lazy. Thunderbirds standby for action.

**Comment by birchb1024** [comment]
Previous attempt on this branch: if-eager-evaluation-#6---Just-warn-on-dot-notation-errors. 
Pull this to workaround error.

**Comment by birchb1024** [comment]
Fixed #6 in 0.3.4

**Comment by lgfausak** [comment]
that does seem like a better approach than just returning null,  i've always liked the fact that an error is returned when a bad reference is made.  is the lazyness everywhere, like if i reference something in a python_eval will that explode?  thank you for the fix, this tool is becoming very useful!
